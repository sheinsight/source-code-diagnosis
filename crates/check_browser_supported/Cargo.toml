[package]
name = "check_browser_supported"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
oxc-browserslist = "1.1.1"
oxc_ast = { workspace = true }
oxc_semantic = { workspace = true }
oxc_span = { workspace = true }
oxc_allocator = { workspace = true }
oxc_parser = { workspace = true }
oxc_syntax = { workspace = true }
serde = { workspace = true }
napi = { workspace = true }
napi-derive = { workspace = true }
log = { workspace = true }
beans = { path = "../beans" }
utils = { path = "../utils" }
anyhow = { workspace = true }
insta = { workspace = true }

[dev-dependencies]
tokio = { workspace = true, features = ["full"] }
