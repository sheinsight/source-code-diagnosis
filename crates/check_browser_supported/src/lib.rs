mod classes;
mod compat;
mod functions;
mod grammar;
mod macros;
mod operators;
mod statements;

pub use compat::{CompatBox, CompatHandler};

use compat_checker::CompatChecker;
use target::Target;

mod browser_versions;
mod compat_checker;
pub mod target;

pub fn check_browser_supported_with_source_code(
  target: Target,
  source_code: String,
  file_path: String,
) -> anyhow::Result<Vec<CompatBox>> {
  let checker = CompatChecker::new(target)?;
  checker.check_source(&source_code, &file_path)
}

pub fn check_browser_supported(
  target: Target,
  args: utils::GlobArgs,
) -> anyhow::Result<Vec<CompatBox>> {
  let checker = CompatChecker::new(target)?;
  checker.check_glob(args)
}
