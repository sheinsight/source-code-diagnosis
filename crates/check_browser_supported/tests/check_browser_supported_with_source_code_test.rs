use check_browser_supported::{
  check_browser_supported_with_source_code, target::Target,
};

#[tokio::test]
async fn test_check_browser_supported_with_source_code() -> anyhow::Result<()> {
  let source_code = r#"
            class MyClass {
                #privateField = 42;
                
                constructor() {
                    console.log(this.#privateField);
                }
            }
            
            new MyClass();
        "#
  .to_string();
  let result = check_browser_supported_with_source_code(
    Target {
      chrome: "40".to_string(),
      firefox: Some("45".to_string()),
      safari: Some("9".to_string()),
      edge: Some("13".to_string()),
      node: Some("6.0.0".to_string()),
      // deno: Some("1.0.0".to_string()),
    },
    source_code,
    "test.ts".to_string(),
  )?;

  insta::assert_debug_snapshot!(result);

  Ok(())
}
