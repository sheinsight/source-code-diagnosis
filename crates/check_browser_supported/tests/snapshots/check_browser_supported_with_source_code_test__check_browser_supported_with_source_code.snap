---
source: crates/check_browser_supported/tests/check_browser_supported_with_source_code_test.rs
expression: result
snapshot_kind: text
---
[
    CompatBox {
        name: "statements.classes",
        compat: Compat {
            name: "statements.classes",
            mdn_url: "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/class",
            description: "The class declaration creates a binding of a new class to a given name.",
            tags: [
                "web-features:class-syntax",
                "web-features:snapshot:ecmascript-2015",
            ],
            support: Support {
                chrome: "49",
                chrome_android: "49",
                firefox: "45",
                firefox_android: "45",
                safari: "10.1",
                safari_ios: "10.1",
                edge: "13",
                node: "6.0.0",
                deno: "1.0.0",
            },
        },
        file_path: "test.ts",
        ast_node: AstNode {
            span: Span {
                start: 13,
                end: 198,
            },
            loc: Location {
                start: Position {
                    line: 2,
                    col: 13,
                },
                end: Position {
                    line: 8,
                    col: 14,
                },
            },
        },
    },
    CompatBox {
        name: "classes.private_class_fields",
        compat: Compat {
            name: "classes.private_class_fields",
            mdn_url: "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties#browser_compatibility",
            description: "Private class fields",
            tags: [
                "web-features:class-syntax",
                "web-features:snapshot:ecmascript-2015",
            ],
            support: Support {
                chrome: "74",
                chrome_android: "74",
                firefox: "90",
                firefox_android: "90",
                safari: "14",
                safari_ios: "14",
                edge: "74",
                node: "12.0.0",
                deno: "1.0.0",
            },
        },
        file_path: "test.ts",
        ast_node: AstNode {
            span: Span {
                start: 45,
                end: 64,
            },
            loc: Location {
                start: Position {
                    line: 3,
                    col: 17,
                },
                end: Position {
                    line: 3,
                    col: 36,
                },
            },
        },
    },
    CompatBox {
        name: "classes.constructor",
        compat: Compat {
            name: "classes.constructor",
            mdn_url: "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/constructor",
            description: "constructor function",
            tags: [
                "web-features:class-syntax",
                "web-features:snapshot:ecmascript-2015",
            ],
            support: Support {
                chrome: "49",
                chrome_android: "49",
                firefox: "45",
                firefox_android: "45",
                safari: "9",
                safari_ios: "9",
                edge: "13",
                node: "6.0.0",
                deno: "1.0.0",
            },
        },
        file_path: "test.ts",
        ast_node: AstNode {
            span: Span {
                start: 98,
                end: 184,
            },
            loc: Location {
                start: Position {
                    line: 5,
                    col: 17,
                },
                end: Position {
                    line: 7,
                    col: 18,
                },
            },
        },
    },
]
